File: benchmarks/pandas/bench_dtype.py
"""Benchmark: pandas dtype access — dtype property, kind, itemsize, numeric checks"""
import json, time
import pandas as pd
import numpy as np

WARMUP = 3
ITERATIONS = 10_000

values = list(range(100))
arr = np.array(values, dtype=np.float64)

for _ in range(WARMUP):
    dt = arr.dtype
    _ = dt.kind
    _ = dt.itemsize
    _ = np.dtype("float64")
    _ = np.result_type(np.dtype("float32"), np.dtype("float64"))
    _ = pd.api.types.is_numeric_dtype(dt)
    _ = pd.api.types.is_float_dtype(dt)
    _ = pd.api.types.is_integer_dtype(dt)

start = time.perf_counter()
for _ in range(ITERATIONS):
    dt = arr.dtype
    _ = dt.kind
    _ = dt.itemsize
    _ = np.dtype("float64")
    _ = np.result_type(np.dtype("float32"), np.dtype("float64"))
    _ = pd.api.types.is_numeric_dtype(dt)
    _ = pd.api.types.is_float_dtype(dt)
    _ = pd.api.types.is_integer_dtype(dt)
total = (time.perf_counter() - start) * 1000
print(json.dumps({"function": "dtype", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_dtype.ts
/**
 * Benchmark: Dtype — singleton lookup, inferFrom, commonType, and property access
 */
import { Dtype } from "../../src/index.js";

const WARMUP = 3;
const ITERATIONS = 10_000;

const values = Array.from({ length: 100 }, (_, i) => i * 1.5);
const mixed = [1, 2.5, "hello", true];

for (let i = 0; i < WARMUP; i++) {
  Dtype.from("float64");
  Dtype.inferFrom(values);
  Dtype.commonType(Dtype.float32, Dtype.float64);
  const dt = Dtype.from("float64");
  dt.isNumeric; dt.isFloat; dt.isInteger; dt.kind; dt.itemsize;
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  Dtype.from("float64");
  Dtype.inferFrom(values);
  Dtype.commonType(Dtype.float32, Dtype.float64);
  const dt = Dtype.from("float64");
  dt.isNumeric; dt.isFloat; dt.isInteger; dt.kind; dt.itemsize;
}
const total = performance.now() - start;
console.log(
  JSON.stringify({
    function: "dtype",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
